1114-i386-slaunch-Add-secure-launch-framework-and-command.patch

```diff
@@ -0,0 +1,891 @@
++
++  err = grub_cpu_is_msr_supported ();
++
++  if (err != GRUB_ERR_NONE)
++    return grub_error (err, N_("MSRs are unsupported"));
++
++  grub_cpuid (0, eax, manufacturer[0], manufacturer[2], manufacturer[1]);
++
++  if (!grub_memcmp (manufacturer, "GenuineIntel", 12))
++    {
++      err = grub_txt_init ();
++
++      if (err != GRUB_ERR_NONE)
++	return err;
++
++      slp = SLP_INTEL_TXT;
++    }
++  else
++    return grub_error (GRUB_ERR_UNKNOWN_DEVICE, N_("CPU is unsupported"));
++
++  return GRUB_ERR_NONE;
++}
++
++static grub_err_t
++grub_cmd_slaunch_module (grub_command_t cmd __attribute__ ((unused)),
++			 int argc, char *argv[])
++{
++  grub_file_t file;
++  grub_ssize_t size;
++
++  if (!argc)
++    return grub_error (GRUB_ERR_BAD_ARGUMENT, N_("filename expected"));
++
++  if (slp == SLP_NONE)
++    return grub_error (GRUB_ERR_BAD_ARGUMENT, N_("secure launch not enabled"));
++
++  grub_errno = GRUB_ERR_NONE;
++
++  file = grub_file_open (argv[0], GRUB_FILE_TYPE_SLAUNCH_MODULE);
++
++  if (file == NULL)
++    return grub_errno;
++
++  size = grub_file_size (file);
++
++  if (!size)
++    {
++      grub_error (GRUB_ERR_BAD_ARGUMENT, N_("file size is zero"));
++      goto fail;
++    }
++
++  slaunch_module = grub_malloc (size);
++
++  if (slaunch_module == NULL)
++    goto fail;
++
++  if (grub_file_read (file, slaunch_module, size) != size)
++    {
++      if (grub_errno == GRUB_ERR_NONE)
++	grub_error (GRUB_ERR_FILE_READ_ERROR, N_("premature end of file: %s"),
++		    argv[0]);
++      goto fail;
++    }
++
++  if (slp == SLP_INTEL_TXT)
++    {
++      if (!grub_txt_is_sinit_acmod (slaunch_module, size))
++	{
++	  grub_error (GRUB_ERR_BAD_FILE_TYPE, N_("it does not look like SINIT ACM"));
++	  goto fail;
++	}
++
++      if (!grub_txt_acmod_match_platform (slaunch_module))
++	{
++	  grub_error (GRUB_ERR_BAD_FILE_TYPE, N_("SINIT ACM does not match platform"));
++	  goto fail;
++	}
++    }
++
++  grub_file_close (file);
++
++  return GRUB_ERR_NONE;
++
++ fail:
++  grub_error_push ();
++
++  grub_free (slaunch_module);
++  grub_file_close (file);
++
++  slaunch_module = NULL;
++
++  grub_error_pop ();
++
++  return grub_errno;
++}
++
++static grub_err_t
++grub_cmd_slaunch_state (grub_command_t cmd __attribute__ ((unused)),
++			int argc __attribute__ ((unused)),
++			char *argv[] __attribute__ ((unused)))
++{
++  if (slp == SLP_NONE)
++    grub_printf ("Secure launcher: Disabled\n");
++  else if (slp == SLP_INTEL_TXT)
++    {
++      grub_printf ("Secure launcher: Intel TXT\n");
++      grub_txt_state_show ();
++    }
++
++  return GRUB_ERR_NONE;
++}
++
++static grub_command_t cmd_slaunch, cmd_slaunch_module, cmd_slaunch_state;
++
++GRUB_MOD_INIT (slaunch)
++{
++  cmd_slaunch = grub_register_command ("slaunch", grub_cmd_slaunch,
++				       NULL, N_("Enable secure launcher"));
++  cmd_slaunch_module = grub_register_command ("slaunch_module", grub_cmd_slaunch_module,
++					      NULL, N_("Secure launcher module command"));
```

Comment:
This description just rephrases the name of the command. How about "Load secure launcher module from file". Depending on how long of a description we want, we could even mention "(chipset ACM for Intel TXT)".

Reply:
I agree with first part, but since we plan to add support for AMD (and perhaps ARM in the future) using the same command for different types of modules, I'd omit the second part.